### PR TITLE
ENH Adds `LayerList` context

### PR DESCRIPTION
**Pull request by lucyleeow**
Adds `LayerList` context. Currently we only have a `Selection[Layer]` context. For the file menu (#4865) we needed `LayerList` because we needed the number of layers present (NOT number of layers selected) and could not obtain this from `Selection[Layer]`. The number of layers present is required for the `enablement` status of the `CommandId.DLG_SAVE_LAYERS` command.

Note this was originally part of #4865 but move to separate PR to aid review and git blame.

# References

ref: https://github.com/napari/napari/pull/4865#discussion_r1292895534

### CODE REVIEW

File: napari/_app_model/context/_layerlist_context.py [hunk 1]
```diff
@@ -12,14 +12,29 @@
...
     from napari.layers import Layer
     from napari.utils.events import Selection
 
     LayerSel = Selection[Layer]
 
 
-def _len(s: LayerSel) -> int:
-    return len(s)
+def _len(layer: Union[LayerSel, LayerList]) -> int:
+    return len(layer)
```
Comment (Carreau) [inline napari/_app_model/context/_layerlist_context.py]:
```suggestion
def _len(layers: Union[LayerSel, LayerList]) -> int:
    return len(layers)
```

?

File: napari/_app_model/context/_layerlist_context.py [hunk 2]
```diff
@@ -134,11 +149,11 @@ def _empty_shapes_layer_selected(s: LayerSel) -> bool:
...
 
-class LayerListContextKeys(ContextNamespace['LayerSel']):
-    """These are the available context keys relating to a LayerList.
+class LayerListSelectionContextKeys(ContextNamespace['LayerSel']):
+    """Available context keys relating to the selection in a LayerList.
 
-    along with default value, a description, and a function to retrieve the
-    current value from layers.selection
+    Consists of a default value, a description, and a function to retrieve the
+    current value from `layers.selection`.
```
Comment (Carreau) [inline napari/_app_model/context/_layerlist_context.py:156]: Thinking at loud, 

This is in a Private file (`_layerlist_context.py`), but this is _technically_ a change of API if LLCK was exposed publicly somewhere, right as we have a rename of LLCK to LLSCK, and a new LLCK with a different behavior. I have no opposition, but I guess it's better to be safe and mak this a breaking change ?

**Review by Carreau — APPROVED** [review]

File: napari/_app_model/context/_layerlist_context.py (hunk 1 above)
Comment (lucyleeow) [reply napari/_app_model/context/_layerlist_context.py]: Though this does highlight the not ideal naming choice below of `s` for Layer selection.

(Changes made)

**Review by kne42 — APPROVED** [review]

File: napari/_app_model/context/_layerlist_context.py (hunk 2 above)
Comment (kne42) [reply napari/_app_model/context/_layerlist_context.py:156]: Since it's not meant to be used or known about by non-napari internal projects, I don't think that label is necessary here and may add noise to the future release notes.

Comment (lucyleeow) [reply napari/_app_model/context/_layerlist_context.py:156]: Is this 'breaking change' marker aimed at plugin developers, or dependencies like npe2 (where there is a lot of overlap with napari developers)?

Comment (kne42) [reply napari/_app_model/context/_layerlist_context.py:156]: I'm not sure exactly @napari/core-devs what are your thoughts? Personally, I view npe2 as internal, treated as if it is an extension of the main napari repo for things like protected functions.

As far as plugin developers, this wouldn't be a breaking change as long as they stick to our public API.

Breaking changes are mainly aimed at people who externally interact with our API since they require a deprecation notice.

Comment (Czaki) [reply napari/_app_model/context/_layerlist_context.py:156]: Do you know about the usage of its outside napari repository? I do not see the conditional option in plugin contributions (based on docs).

Comment (lucyleeow) [reply napari/_app_model/context/_layerlist_context.py:156]: >  aimed at people who externally interact with our API since they require a deprecation notice.

Sorry my question was more, apart from plugin devs, who else external interacts with our API?

> Do you know about the usage of its outside napari repository? I do not see the conditional option in plugin contributions (based on docs).

No but it is planned to be made public as part of nap 6 (#6174)

Comment (DragaDoncila) [reply napari/_app_model/context/_layerlist_context.py:156]: Given that this is in a private file and not exposed to plugin developers yet, I have no qualms about changing the name without a deprecation strategy here. A code search for `LayerListContextKeys` only revealed napari forks - although @Czaki there were a couple of mentions in `napari-release-tools` but I think they're just docs?

Comment (DragaDoncila) [reply napari/_app_model/context/_layerlist_context.py:156]: > Sorry my question was more, apart from plugin devs, who else external interacts with our API?

In theory, any napari user **could** be writing scripts that make use of these context keys in some way, including making on the fly contributions, but it's exceedingly unlikely that they're doing it right now.

**Review by DragaDoncila — APPROVED** [review]
LGTM thanks @lucyleeow! As I mentioned, I'm not concerned about the name change in this case.

**Review by Carreau — APPROVED** [review]
I slightly edited the description, to cleanup the text from the template, and removed any @-mention, as the description will become the commit text, and any @-mention in it will trigger everytime someone push the merge commit somewhere.

**Comment by DragaDoncila** [comment]
Thanks folks I think all comments have been addressed and we have three approvals, so I'm merging now!

File: napari/components/layerlist.py [hunk 3]
```diff
@@ -59,21 +59,43 @@ def __init__(self, data=()) -> None:
...
 
         self._ctx = create_context(self)
         if self._ctx is not None:  # happens during Viewer type creation
             self._ctx_keys = LayerListContextKeys(self._ctx)
-
-            self.selection.events.changed.connect(self._ctx_keys.update)
+            self.events.inserted.connect(self._ctx_keys.update)
+            self.events.removed.connect(self._ctx_keys.update)
+
+        self._selection_ctx = create_context(self)
```
Comment (lucyleeow) [inline napari/components/layerlist.py:89]: @tlambert03 this may be a silly question but when I create another `Context` using `self` here, would the `id` of `self` be the same? So when [`create_context`](https://github.com/pyapp-kit/app-model/blob/4d796c6390c2c4141f370f93a3162731e0d940e5/src/app_model/expressions/_context.py#L80) does:

```python
_OBJ_TO_CONTEXT[obj_id] = new_context
```

am I replacing the old dict item?

Comment (tlambert03) [reply napari/components/layerlist.py:89]: nope, each `LayerList` instance will have a unique `id()`; more generally, all python objects (that haven't been garbage collected)  have a unique id... anytime `object_a is not object_b`, it means they have a different id.

Comment (lucyleeow) [reply napari/components/layerlist.py:89]: Sorry I should have been explicit in that I use `create_context` twice in the init, i.e. in the same instance. But thinking about it I think instead of:

```python
        self._ctx = create_context(self)
        if self._ctx is not None:  # happens during Viewer type creation
            self._ctx_keys = LayerListContextKeys(self._ctx)
            self.events.inserted.connect(self._ctx_keys.update)
            self.events.removed.connect(self._ctx_keys.update)

        self._selection_ctx = create_context(self)
        if (
            self._selection_ctx is not None
        ):  # happens during Viewer type creation
            self._selection_ctx_keys = LayerListSelectionContextKeys(
                self._selection_ctx
            )
```

it should just be:

```python
        self._ctx = create_context(self)
        if self._ctx is not None:  # happens during Viewer type creation
            self._ctx_keys = LayerListContextKeys(self._ctx)
            self.events.inserted.connect(self._ctx_keys.update)
            self.events.removed.connect(self._ctx_keys.update)

            self._selection_ctx_keys = LayerListSelectionContextKeys(
                self._ctx
            )
```

i.e. I can just use the same context to create the `LayerListContextKeys` and `LayerListSelectionContextKeys`.

Is that right?

Comment (tlambert03) [reply napari/components/layerlist.py:89]: ohh... I see, you're making a whole new context for the same object...  why?  

yes that would make any future calls to `get_context` return only the second one.  it definitely wasn't the intended usage of `create_context`. (I probably would/should have thrown an exception).

Comment (lucyleeow) [reply napari/components/layerlist.py:89]: > new context for the same object... why?

I didn't know what I was doing 😅

Comment (tlambert03) [reply napari/components/layerlist.py:89]: Aw, well, no biggie.  Are you calling get_context anywhere on the layerlist? If not, no harm done, and if so, easily changed

Comment (lucyleeow) [reply napari/components/layerlist.py:89]: >  Are you calling get_context anywhere on the layerlist?

Yes I do, I manually up `update_from_context` for all the menubar menus:

https://github.com/napari/napari/blob/98c2e0f39982d46aebc70acf5a32dd2beaa23df4/napari/_qt/qt_main_window.py#L817

which is used here:

https://github.com/napari/napari/blob/98c2e0f39982d46aebc70acf5a32dd2beaa23df4/napari/_qt/qt_main_window.py#L868-L874

I thought I checked and this (the next 'selection' context I created) still worked...?

Does my double up in calling `get_context` create a different `Context` the second time round? Or is the double up just redundant but doesn't change anything?

Comment (tlambert03) [reply napari/components/layerlist.py:89]: ok... looked at this a bit more.   It does work as you have it here, but I'd say the fact that it works is a bit surprising, and probably shouldn't be depended on as a pattern.  It works because when you call `self._selection_ctx = create_context(self)` (after having already created a context for `self`), you end up hitting [this branch in app_model's `create_context`](https://github.com/pyapp-kit/app-model/blob/81b245baa45cb3bc7e1afd81d3af06dfe6fa1ca9/src/app_model/expressions/_context.py#L134-L140):

```python
            # FIXME: this might be a bit napari "magic"
            # it also assumes someone uses "self" as the first argument
            if "self" in frame.f_locals:
                _ctx = _OBJ_TO_CONTEXT.get(id(frame.f_locals["self"]))
                if _ctx is not None:
                    parent = _ctx
```

What that chunk does is it asks: "have we already created a context for this object? if so, then scope the context we are about to create off of this existing context"

Meaning: the new context you get back `self._selection_ctx` is a `ChainMap` that has the previously created context `self._ctx`.  (and so any keys in the original self._ctx can also be accessed from self._selection_ctx) You can see this by running:

```python
        self._selection_ctx = create_context(self)
        print("has self._ctx?:", self._ctx in self._selection_ctx.maps)
```

(If that weren't the case, then nothing would be "responding" to changes in the `LayerListContextKeys.num_layers`, which, at the moment, appears to only be the `napari.window.file.save_layers_dialog` Action... so, without this chain map magic here, your `Save All Layers...` menu item would never be enabled).

in summary, it's working fine, but probably only by accident.

Comment (lucyleeow) [reply napari/components/layerlist.py:89]: Thanks, Draga and I had a dig and thought the same!

I guess it would be hard to warn if a context has been created with the same object id before without a context registry as you'd need to keep track of ids.

Comment (DragaDoncila) [reply napari/components/layerlist.py:89]: > I guess it would be hard to warn if a context has been created with the same object id before without a context registry as you'd need to keep track of ids.

I mean we track it in `_OBJ_TO_CONTEXT` so we could just check if the id is already a key in the dictionary, no?

Comment (lucyleeow) [reply napari/components/layerlist.py:89]: :facepalm: omg yes of course, would be trivial.
